File: May 2022/May_12.cpp
class Solution {
public:
    void helper(unordered_map<int, int> &mp, vector<vector<int> > &ans, vector<int> &v){
        if(mp.empty()){
            vector<int> tmp;
            for(auto &i : v)
                tmp.push_back(i);
            ans.push_back(tmp);
            return;
        }
        
        unordered_map<int, int> nmp;
        for(auto &p : mp)
            nmp[p.first] = p.second;
        
        for(auto &p : mp){
            if(p.second == 1){
                nmp.erase(p.first);
                v.push_back(p.first);
                helper(nmp, ans, v);
                v.pop_back();
                nmp[p.first] = 1;
            } else{
                nmp[p.first] = p.second - 1;
                v.push_back(p.first);
                helper(nmp, ans, v);
                v.pop_back();
                nmp[p.first] = p.second;
            }
        }
    }
    
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int> > ans;
        unordered_map<int, int> mp;
        for(auto &i : nums)
            mp[i]++;
        vector<int> v;
        helper(mp, ans, v);
        return ans;
    }
};

File: May 2022/May_12.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        list = new ArrayList<>();
        Map<Integer, Integer> map = new HashMap<>();
        for(int i : nums)
            map.put(i, map.getOrDefault(i, 0) + 1);
        helper(nums.length, nums, new ArrayList<>(), map);
        return list;
    }
    
    List<List<Integer>> list;
    
    private void helper(int n, int[] nums, List<Integer> al, Map<Integer, Integer> map){
        if(map.isEmpty()){
            list.add(new ArrayList<>(al));
            return;
        }
        
        Map<Integer, Integer> nmap = new HashMap<>(map);
        for(int i : map.keySet()){
            int f = map.get(i);
            if(f == 1){
                nmap.remove(i);
                al.add(i);
                helper(n, nums, al, nmap);
                al.remove(al.size() - 1);
                nmap.put(i, 1);
            } else{
                nmap.put(i, f - 1);
                al.add(i);
                helper(n, nums, al, nmap);
                al.remove(al.size() - 1);
                nmap.put(i, f);
            }
        }
    }
}
